Src/ReadBarometer.c

```diff
@@ -6,127 +15,256 @@
+		//
+		// Read Digital Temperature (D2)
+		//
 
-        HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
-        pressureReading += dataInBuffer << 16;
+		// Tell the barometer to convert the temperature to a digital value with an OSR of 512
+		HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_RESET);
+		HAL_SPI_Transmit(&hspi2, &ADC_D2_512_CONV_CMD, CMD_SIZE, CMD_TIMEOUT);
+		HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_SET);
 
-        HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
-        pressureReading += dataInBuffer << 8;
+		osDelay(2); // 1.17ms max conversion time for OSR 512
 
-        HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
-        pressureReading += dataInBuffer;
+		// Read the temperature value
 
-        HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_SET);
+		// Chip select to start a new command
+		HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_RESET);
 
-        // Read D2 temperature
-        HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_RESET);
-        HAL_SPI_Transmit(&hspi2, &ADC_D2_512_CONV_CMD, CMD_SIZE, CMD_TIMEOUT);
-        HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_SET);
+		// Tell the chip we want to read from the adc
+		HAL_SPI_Transmit(&hspi2, &ADC_READ_CMD, CMD_SIZE, CMD_TIMEOUT);
 
-        osDelay(2); // 1.17ms max conversion time for OSR 512
+		// Read the first byte (bits 23-16)
+		HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
+		temperatureReading = dataInBuffer << 16;
 
-        HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_RESET);
-        HAL_SPI_Transmit(&hspi2, &ADC_READ_CMD, CMD_SIZE, CMD_TIMEOUT);
+		// Read the second byte (bits 15-8)
+		HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
+		temperatureReading += dataInBuffer << 8;
 
-        temperatureReading = 0;
+		// Read the third byte (bits 7-0)
+		HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
+		temperatureReading += dataInBuffer;
 
-        HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
-        temperatureReading += dataInBuffer << 16;
+		// Release chip select to end the command
+		HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_SET);
 
-        HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
-        temperatureReading += dataInBuffer << 8;
+		//
+		// Calculate First-Order Temperature and Pressure
+		//
 
-        HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
-        temperatureReading += dataInBuffer;
+		// Equations provided by MS5607-02BA03 data sheet
 
-        HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_SET);
+		int32_t dT		= temperatureReading - (c5Tref << 8);
```

Comment:
If we do not use doubles, we are rounding intermediate values that are used in later calculations. This can reduce the accuracy of the calculations, but I am not sure if it will cause a problem.

Reply:
Okay I missed that in the datasheet. In that case, it makes sense to use signed integers just like the datasheet does.